File: foundational_design_patterns/6_planning/src/planning_hiplan.py
"""
Planning Pattern: HiPlan-style hierarchical decomposition

Anchor scenario: onboarding automation for a new enterprise customer.

This script demonstrates:
- high-level milestones
- sub-plan generation per milestone
- milestone-by-milestone execution summaries
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass, field

from pathlib import Path

ROOT_DIR = next(parent for parent in Path(__file__).resolve().parents if (parent / "ssl_fix.py").exists())
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from repo_support import configure_example, get_default_model

configure_example(__file__)

from langchain_openai import ChatOpenAI


@dataclass
class Milestone:
    """HiPlan milestone with generated subtasks."""

    name: str
    goal: str
    subtasks: list[str] = field(default_factory=list)


llm = ChatOpenAI(model=get_default_model(), temperature=0)


def generate_milestones(objective: str) -> list[Milestone]:
    prompt = f"""
Create 5 milestones for the objective below.
Return each milestone as: name | goal

Objective:
{objective}
"""
    lines = llm.invoke(prompt).content.strip().splitlines()

    milestones: list[Milestone] = []
    for line in lines:
        if "|" not in line:
            continue
        left, right = line.split("|", 1)
        milestones.append(Milestone(name=left.strip("-• ").strip(), goal=right.strip()))

    return milestones[:5]


def generate_subtasks(milestone: Milestone, objective: str) -> Milestone:
    prompt = f"""
You are planning execution details for one milestone.
Objective: {objective}
Milestone: {milestone.name}
Goal: {milestone.goal}

Return 4 atomic subtasks, one per line.
"""
    lines = llm.invoke(prompt).content.strip().splitlines()
    milestone.subtasks = [line.strip("-• ").strip() for line in lines if line.strip()][:4]
    return milestone


def execute_subtasks(milestone: Milestone) -> list[str]:
    logs: list[str] = []
    for subtask in milestone.subtasks:
        prompt = f"""
Simulate execution of this onboarding subtask and return one short status sentence.
Subtask: {subtask}
"""
        status = llm.invoke(prompt).content.strip()
        logs.append(f"{subtask} -> {status}")
    return logs


def main() -> None:
    print("\n" + "=" * 80)
    print("PLANNING PATTERN — HIPLAN HIERARCHICAL DECOMPOSITION")
    print("=" * 80)

    if not os.getenv("OPENAI_API_KEY"):
        print("❌ OPENAI_API_KEY is required for this example.")
        return

    objective = (
        "Onboard a new enterprise customer for our SaaS platform, including security review, "
        "identity integration, data migration, training, and go-live support."
    )

    milestones = generate_milestones(objective)
    if not milestones:
        print("⚠️ Could not generate milestones.")
        return

    print("\n🏁 Milestones:")
    for idx, ms in enumerate(milestones, 1):
        print(f"  {idx}. {ms.name} — {ms.goal}")

    print("\n🔧 Generating subtasks and execution logs...")
    for ms in milestones:
        generate_subtasks(ms, objective)
        logs = execute_subtasks(ms)

        print(f"\n📌 {ms.name}")
        for sub in ms.subtasks:
            print(f"   - {sub}")
        print("   Execution:")
        for log in logs:
            print(f"     • {log}")


if __name__ == "__main__":
    main()

File: foundational_design_patterns/6_planning/typescript/src/planning_hiplan.ts
/**
 * Planning Pattern: HiPlan-style hierarchical decomposition
 * — TypeScript port of src/planning_hiplan.py.
 *
 * Anchor scenario: onboarding automation for a new enterprise customer.
 *
 * This script demonstrates:
 *   - high-level milestones
 *   - sub-plan generation per milestone
 *   - milestone-by-milestone execution summaries
 */

import { ChatOpenAI } from "@langchain/openai";

const MODEL = process.env.OPENAI_MODEL ?? "gpt-4o-mini";

// ---------------------------------------------------------------------------
// HiPlan milestone with generated subtasks.
// ---------------------------------------------------------------------------

export interface Milestone {
  name: string;
  goal: string;
  subtasks: string[];
}

// ---------------------------------------------------------------------------
// LLM (lazy-init so smoke tests can import without OPENAI_API_KEY).
// ---------------------------------------------------------------------------

let _llm: ChatOpenAI | undefined;
function llm(): ChatOpenAI {
  if (!_llm) {
    _llm = new ChatOpenAI({ model: MODEL, temperature: 0 });
  }
  return _llm;
}

function contentToString(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((part) =>
        typeof part === "string"
          ? part
          : typeof part === "object" && part !== null && "text" in part
            ? String((part as { text: unknown }).text)
            : "",
      )
      .join("");
  }
  return String(content);
}

// ---------------------------------------------------------------------------
// Milestone / subtask / execution generators
// ---------------------------------------------------------------------------

export async function generateMilestones(objective: string): Promise<Milestone[]> {
  const prompt = `
Create 5 milestones for the objective below.
Return each milestone as: name | goal

Objective:
${objective}
`;
  const response = await llm().invoke(prompt);
  const lines = contentToString(response.content).trim().split(/\r?\n/);

  const milestones: Milestone[] = [];
  for (const line of lines) {
    if (!line.includes("|")) continue;
    const sep = line.indexOf("|");
    const left = line.slice(0, sep).replace(/^[-•\s]+/, "").trim();
    const right = line.slice(sep + 1).trim();
    milestones.push({ name: left, goal: right, subtasks: [] });
  }

  return milestones.slice(0, 5);
}

export async function generateSubtasks(
  milestone: Milestone,
  objective: string,
): Promise<Milestone> {
  const prompt = `
You are planning execution details for one milestone.
Objective: ${objective}
Milestone: ${milestone.name}
Goal: ${milestone.goal}

Return 4 atomic subtasks, one per line.
`;
  const response = await llm().invoke(prompt);
  const lines = contentToString(response.content).trim().split(/\r?\n/);
  milestone.subtasks = lines
    .map((line) => line.replace(/^[-•\s]+/, "").trim())
    .filter((line) => line.length > 0)
    .slice(0, 4);
  return milestone;
}

export async function executeSubtasks(milestone: Milestone): Promise<string[]> {
  const logs: string[] = [];
  for (const subtask of milestone.subtasks) {
    const prompt = `
Simulate execution of this onboarding subtask and return one short status sentence.
Subtask: ${subtask}
`;
    const response = await llm().invoke(prompt);
    const status = contentToString(response.content).trim();
    logs.push(`${subtask} -> ${status}`);
  }
  return logs;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  console.log("\n" + "=".repeat(80));
  console.log("PLANNING PATTERN — HIPLAN HIERARCHICAL DECOMPOSITION");
  console.log("=".repeat(80));

  if (!process.env.OPENAI_API_KEY) {
    console.log("❌ OPENAI_API_KEY is required for this example.");
    return;
  }

  const objective =
    "Onboard a new enterprise customer for our SaaS platform, including security review, " +
    "identity integration, data migration, training, and go-live support.";

  const milestones = await generateMilestones(objective);
  if (milestones.length === 0) {
    console.log("⚠️ Could not generate milestones.");
    return;
  }

  console.log("\n🏁 Milestones:");
  milestones.forEach((ms, idx) => {
    console.log(`  ${idx + 1}. ${ms.name} — ${ms.goal}`);
  });

  console.log("\n🔧 Generating subtasks and execution logs...");
  for (const ms of milestones) {
    await generateSubtasks(ms, objective);
    const logs = await executeSubtasks(ms);

    console.log(`\n📌 ${ms.name}`);
    for (const sub of ms.subtasks) {
      console.log(`   - ${sub}`);
    }
    console.log("   Execution:");
    for (const log of logs) {
      console.log(`     • ${log}`);
    }
  }
}

if (import.meta.main) {
  await main();
}
